File: backendfastapi/services/agent_service.py
from typing import List, Optional, Dict, Any
from datetime import datetime
import uuid
from lib.mongodb import get_database
from models.agent import Agent, AgentTemplate, AgentTool, AgentToolType
from bson import ObjectId

class AgentService:
    def __init__(self):
        self._db = None
        self._agents_collection = None
        self._templates_collection = None
    
    @property
    def db(self):
        if self._db is None:
            self._db = get_database()
        return self._db
    
    @property
    def agents_collection(self):
        if self._agents_collection is None:
            self._agents_collection = self.db.agents
        return self._agents_collection
    
    @property
    def templates_collection(self):
        if self._templates_collection is None:
            self._templates_collection = self.db.agent_templates
        return self._templates_collection
    
    def get_default_templates(self):
        """Get default agent templates"""
        return [
            {
                "_id": "programming-assistant",
                "name": "Programming Assistant",
                "description": "Expert in programming languages, debugging, and code review",
                "category": "Development",
                "icon": "💻",
                "systemPrompt": "You are an expert programming assistant. Help users with coding questions, debugging, code review, and software development best practices. Provide clear, practical solutions with examples. Always focus on answering the current user's question. Use chat history as context to provide better responses, but do not repeat or respond to previous questions in the history.",
                "recommendedTools": ["web_search", "calculator"],
                "recommendedCollections": ["programming-docs", "api-documentation"],
                "tags": ["programming", "coding", "development"]
            },
            {
                "_id": "academic-tutor",
                "name": "Academic Tutor",
                "description": "Specialized in academic subjects and research assistance",
                "category": "Education",
                "icon": "🎓",
                "systemPrompt": "You are an academic tutor. Provide clear explanations, help students understand complex concepts, and assist with research. Use evidence-based information and cite sources when appropriate. Always focus on answering the current user's question. Use chat history as context to provide better responses, but do not repeat or respond to previous questions in the history.",
                "recommendedTools": ["web_search"],
                "recommendedCollections": ["academic-papers", "textbooks", "research-data"],
                "tags": ["education", "academic", "research"]
            },
            {
                "_id": "writing-assistant",
                "name": "Writing Assistant",
                "description": "Professional writing and content creation support",
                "category": "Content",
                "icon": "✍️",
                "systemPrompt": "You are a professional writing assistant. Help users with content creation, editing, proofreading, and improving writing style. Provide constructive feedback and suggestions. Always focus on answering the current user's question. Use chat history as context to provide better responses, but do not repeat or respond to previous questions in the history.",
                "recommendedTools": ["web_search"],
                "recommendedCollections": ["writing-guides", "style-manuals"],
                "tags": ["writing", "content", "editing"]
            }
        ]

    async def get_all_agents(self, user_id: Optional[str] = None) -> List[Agent]:
        """Get all agents, optionally filtered by user"""
        try:
            query = {}
            if user_id:
                query = {"$or": [
                    {"createdBy": user_id},
                    {"isPublic": True}
                ]}
            
            agents_data = await self.agents_collection.find(query).to_list(length=None)
            
            # Convert ObjectId to string
            for agent in agents_data:
                agent["id"] = str(agent.pop("_id", ""))
                
            return [Agent(**agent) for agent in agents_data]
        except Exception as e:
            print(f"Error fetching agents: {e}")
            return []

    async def get_agent_by_id(self, agent_id: str) -> Optional[Agent]:
        """Get agent by ID"""
        try:
            # Handle special case for default agent ID
            if agent_id == '000000000000000000000001':
                # Return a default agent
                return Agent(
                    id='000000000000000000000001',
                    name='General Assistant',
                    description='A helpful AI assistant for general questions and tasks',
                    systemPrompt='You are a helpful AI assistant. Provide clear, accurate, and helpful responses to user questions. Always focus on answering the current user\'s question. Use chat history as context to provide better responses, but do not repeat or respond to previous questions in the history.',
                    modelId='anthropic.claude-3-5-sonnet-20240620-v1:0',
                    collectionNames=[],
                    tools=[],
                    temperature=0.7,
                    maxTokens=4000,
                    isPublic=True,
                    tags=['general', 'assistant'],
                    createdBy='system',
                    createdAt=datetime.utcnow(),
                    updatedAt=datetime.utcnow(),
                    usageCount=0,
                    rating=0.0
                )
            
            # Validate ObjectId format
            if not ObjectId.is_valid(agent_id):
                print(f"Invalid ObjectId format: {agent_id}")
                return None
                
            agent_data = await self.agents_collection.find_one({"_id": ObjectId(agent_id)})
            if agent_data:
                agent_data["id"] = str(agent_data.pop("_id"))
                return Agent(**agent_data)
            return None
        except Exception as e:
            print(f"Error fetching agent {agent_id}: {e}")
            return None

    async def create_agent(self, agent_data: Dict[str, Any]) -> Agent:
        """Create a new agent"""
        try:
            # Prepare agent document
            agent_doc = {
                "_id": ObjectId(),
                "name": agent_data["name"],
                "description": agent_data.get("description", ""),
                "systemPrompt": agent_data.get("systemPrompt", ""),
                "modelId": agent_data["modelId"],
                "collectionNames": agent_data.get("collectionNames", []),
                "tools": agent_data.get("tools", []),
                "temperature": agent_data.get("temperature", 0.7),
                "maxTokens": agent_data.get("maxTokens", 4000),
                "isPublic": agent_data.get("isPublic", False),
                "tags": agent_data.get("tags", []),
                "createdBy": agent_data["createdBy"],
                "createdAt": datetime.utcnow(),
                "updatedAt": datetime.utcnow(),
                "usageCount": 0,
                "rating": 0.0
            }
            
            result = await self.agents_collection.insert_one(agent_doc)
            
            # Return created agent
            agent_doc["id"] = str(agent_doc.pop("_id"))
            return Agent(**agent_doc)
            
        except Exception as e:
            print(f"Error creating agent: {e}")
            raise ValueError(f"Failed to create agent: {str(e)}")

    async def update_agent(self, agent_id: str, updates: Dict[str, Any]) -> Optional[Agent]:
        """Update an existing agent"""
        try:
            updates["updatedAt"] = datetime.utcnow()
            
            result = await self.agents_collection.update_one(
                {"_id": ObjectId(agent_id)},
                {"$set": updates}
            )
            
            if result.matched_count > 0:
                return await self.get_agent_by_id(agent_id)
            
            return None
        except Exception as e:
            print(f"Error updating agent {agent_id}: {e}")
            return None

    async def delete_agent(self, agent_id: str) -> bool:
        """Delete an agent"""
        try:
            result = await self.agents_collection.delete_one({"_id": ObjectId(agent_id)})
            return result.deleted_count > 0
        except Exception as e:
            print(f"Error deleting agent {agent_id}: {e}")
            return False

    async def get_agent_templates(self) -> List[AgentTemplate]:
        """Get all agent templates"""
        try:
            # Check if templates exist in database
            templates_data = await self.templates_collection.find().to_list(length=None)
            
            # If no templates in database, initialize with defaults
            if not templates_data:
                default_templates = self.get_default_templates()
                await self.templates_collection.insert_many(default_templates)
                templates_data = default_templates
            
            # Convert to AgentTemplate objects
            templates = []
            for template in templates_data:
                template["id"] = template.get("_id", template.get("id"))
                if "_id" in template:
                    del template["_id"]
                templates.append(AgentTemplate(**template))
            
            return templates
        except Exception as e:
            print(f"Error fetching templates: {e}")
            return []

    async def create_agent_from_template(self, template_id: str, customizations: Dict[str, Any]) -> Agent:
        """Create an agent from a template"""
        try:
            # Get template
            template_data = await self.templates_collection.find_one({"_id": template_id})
            if not template_data:
                # Try from default templates
                default_templates = self.get_default_templates()
                template_data = next((t for t in default_templates if t["_id"] == template_id), None)
                if not template_data:
                    raise ValueError("Template not found")
            
            # Create agent data from template
            agent_data = {
                "name": customizations.get("name", template_data["name"]),
                "description": customizations.get("description", template_data["description"]),
                "systemPrompt": customizations.get("systemPrompt", template_data["systemPrompt"]),
                "modelId": customizations.get("modelId", "anthropic.claude-3-5-sonnet-20240620-v1:0"),
                "collectionNames": customizations.get("collectionNames", template_data.get("recommendedCollections", [])),
                "tools": customizations.get("tools", self._create_tools_from_recommendations(template_data.get("recommendedTools", []))),
                "temperature": customizations.get("temperature", 0.7),
                "maxTokens": customizations.get("maxTokens", 4000),
                "isPublic": customizations.get("isPublic", False),
                "tags": customizations.get("tags", template_data.get("tags", [])),
                "createdBy": customizations["createdBy"]
            }
            
            return await self.create_agent(agent_data)
            
        except Exception as e:
            print(f"Error creating agent from template: {e}")
            raise ValueError(f"Failed to create agent from template: {str(e)}")

    def _create_tools_from_recommendations(self, recommended_tools: List[str]) -> List[Dict[str, Any]]:
        """Create tool objects from recommended tool names"""
        tools = []
        for tool_type in recommended_tools:
            tool = {
                "id": tool_type,
                "name": tool_type.replace('_', ' ').title(),
                "description": f"{tool_type} tool",
                "type": tool_type,
                "config": {},
                "enabled": True
            }
            tools.append(tool)
        return tools

    async def increment_usage_count(self, agent_id: str) -> None:
        """Increment the usage count for an agent"""
        try:
            await self.agents_collection.update_one(
                {"_id": ObjectId(agent_id)},
                {"$inc": {"usageCount": 1}}
            )
        except Exception as e:
            print(f"Error incrementing usage count for agent {agent_id}: {e}")

    async def update_agent_rating(self, agent_id: str, rating: float) -> None:
        """Update the rating for an agent"""
        try:
            await self.agents_collection.update_one(
                {"_id": ObjectId(agent_id)},
                {"$set": {"rating": rating, "updatedAt": datetime.utcnow()}}
            )
        except Exception as e:
            print(f"Error updating rating for agent {agent_id}: {e}")

# Create singleton instance
agent_service = AgentService() 

File: backend/src/services/agentService.ts
import { AgentModel, AgentTemplateModel, Agent, AgentTemplate, AgentTool, AgentToolType } from '../models/agent';
import { v4 as uuidv4 } from 'uuid';

export class AgentService {
  constructor() {
    console.log('✅ Agent service initialized');
  }

  public async getAllAgents(userId?: string): Promise<Agent[]> {
    try {
      let query = {};
      if (userId) {
        query = {
          $or: [
            { createdBy: userId },
            { isPublic: true }
          ]
        };
      }
      
      const agents = await AgentModel.find(query).exec();
      return agents;
    } catch (error) {
      console.error('Error fetching agents:', error);
      return [];
    }
  }

  public async getAgentById(agentId: string): Promise<Agent | null> {
    try {
      // Handle special case for default agent ID
      if (agentId === '000000000000000000000001') {
        return this.getDefaultAgent();
      }
      
      const agent = await AgentModel.findById(agentId);
      return agent;
    } catch (error) {
      console.error(`Error fetching agent ${agentId}:`, error);
      return null;
    }
  }

  private getDefaultAgent(): Agent {
    return new AgentModel({
      name: 'General Assistant',
      description: 'A helpful AI assistant for general questions and tasks',
      systemPrompt: 'You are a helpful AI assistant. Provide clear, accurate, and helpful responses to user questions. Always focus on answering the current user\'s question. Use chat history as context to provide better responses, but do not repeat or respond to previous questions in the history.',
      modelId: 'anthropic.claude-3-5-sonnet-20240620-v1:0',
      collectionNames: [],
      tools: [
        { id: uuidv4(), name: 'Web Search', description: 'Search the web for current information', type: AgentToolType.WEB_SEARCH, config: {}, enabled: true },
        { id: uuidv4(), name: 'Calculator', description: 'Perform mathematical calculations', type: AgentToolType.CALCULATOR, config: {}, enabled: true },
        { id: uuidv4(), name: 'Current Date', description: 'Get the current date and time', type: AgentToolType.CURRENT_DATE, config: {}, enabled: true },
        { id: uuidv4(), name: 'Memory Search', description: 'Search through chat memory for relevant context', type: AgentToolType.MEMORY_SEARCH, config: {}, enabled: true },
        { id: uuidv4(), name: 'Memory Embed', description: 'Embed new message into chat memory', type: AgentToolType.MEMORY_EMBED, config: {}, enabled: true }
      ],
      temperature: 0.7,
      maxTokens: 4000,
      isPublic: true,
      tags: ['general', 'assistant'],
      createdBy: 'system',
      createdAt: new Date(),
      updatedAt: new Date(),
      usageCount: 0,
      rating: 0.0
    });
  }

  public async createAgent(agentData: any): Promise<Agent> {
    try {
      const agent = new AgentModel({
        name: agentData.name,
        description: agentData.description || '',
        systemPrompt: agentData.systemPrompt || '',
        modelId: agentData.modelId,
        collectionNames: agentData.collectionNames || [],
        tools: agentData.tools || [],
        temperature: agentData.temperature || 0.7,
        maxTokens: agentData.maxTokens || 4000,
        isPublic: agentData.isPublic || false,
        tags: agentData.tags || [],
        createdBy: agentData.createdBy,
        createdAt: new Date(),
        updatedAt: new Date(),
        usageCount: 0,
        rating: 0.0
      });

      await agent.save();
      console.log(`✅ Created agent: ${agent.name}`);
      return agent;
    } catch (error) {
      console.error('Error creating agent:', error);
      throw new Error(`Failed to create agent: ${error}`);
    }
  }

  public async updateAgent(agentId: string, updates: any): Promise<Agent | null> {
    try {
      updates.updatedAt = new Date();
      
      const agent = await AgentModel.findByIdAndUpdate(
        agentId,
        { $set: updates },
        { new: true }
      );
      
      if (agent) {
        console.log(`✅ Updated agent: ${agent.name}`);
      }
      
      return agent;
    } catch (error) {
      console.error(`Error updating agent ${agentId}:`, error);
      return null;
    }
  }

  public async deleteAgent(agentId: string): Promise<boolean> {
    try {
      const result = await AgentModel.findByIdAndDelete(agentId);
      const success = !!result;
      
      if (success) {
        console.log(`🗑️ Deleted agent: ${result?.name}`);
      }
      
      return success;
    } catch (error) {
      console.error(`Error deleting agent ${agentId}:`, error);
      return false;
    }
  }

  public async getAgentTemplates(): Promise<AgentTemplate[]> {
    try {
      // Check if templates exist in database
      let templates = await AgentTemplateModel.find().exec();
      
      // If no templates in database, initialize with defaults
      if (templates.length === 0) {
        const defaultTemplates = this.getDefaultTemplates();
        await AgentTemplateModel.insertMany(defaultTemplates);
        templates = await AgentTemplateModel.find().exec();
      }
      
      return templates;
    } catch (error) {
      console.error('Error fetching agent templates:', error);
      return [];
    }
  }

  private getDefaultTemplates(): any[] {
    return [
      {
        name: 'Programming Assistant',
        description: 'Expert in programming languages, debugging, and code review',
        category: 'Development',
        icon: '💻',
        systemPrompt: 'You are an expert programming assistant. Help users with coding questions, debugging, code review, and software development best practices. Provide clear, practical solutions with examples. Always focus on answering the current user\'s question. Use chat history as context to provide better responses, but do not repeat or respond to previous questions in the history.',
        recommendedTools: ['web_search', 'calculator'],
        recommendedCollections: ['programming-docs', 'api-documentation'],
        tags: ['programming', 'coding', 'development']
      },
      {
        name: 'Academic Tutor',
        description: 'Specialized in academic subjects and research assistance',
        category: 'Education',
        icon: '🎓',
        systemPrompt: 'You are an academic tutor. Provide clear explanations, help students understand complex concepts, and assist with research. Use evidence-based information and cite sources when appropriate. Always focus on answering the current user\'s question. Use chat history as context to provide better responses, but do not repeat or respond to previous questions in the history.',
        recommendedTools: ['web_search'],
        recommendedCollections: ['academic-papers', 'textbooks', 'research-data'],
        tags: ['education', 'academic', 'research']
      },
      {
        name: 'Writing Assistant',
        description: 'Professional writing and content creation support',
        category: 'Content',
        icon: '✍️',
        systemPrompt: 'You are a professional writing assistant. Help users with content creation, editing, proofreading, and improving writing style. Provide constructive feedback and suggestions. Always focus on answering the current user\'s question. Use chat history as context to provide better responses, but do not repeat or respond to previous questions in the history.',
        recommendedTools: ['web_search'],
        recommendedCollections: ['writing-guides', 'style-manuals'],
        tags: ['writing', 'content', 'editing']
      }
    ];
  }

  public async createAgentFromTemplate(templateId: string, customizations: any): Promise<Agent> {
    try {
      const template = await AgentTemplateModel.findById(templateId);
      if (!template) {
        throw new Error('Template not found');
      }

      const tools = this.createToolsFromRecommendations(template.recommendedTools);
      
      const agentData = {
        name: customizations.name || template.name,
        description: customizations.description || template.description,
        systemPrompt: customizations.systemPrompt || template.systemPrompt,
        modelId: customizations.modelId || 'anthropic.claude-3-5-sonnet-20240620-v1:0',
        collectionNames: customizations.collectionNames || template.recommendedCollections,
        tools,
        temperature: customizations.temperature || 0.7,
        maxTokens: customizations.maxTokens || 4000,
        isPublic: customizations.isPublic || false,
        tags: customizations.tags || template.tags,
        createdBy: customizations.createdBy
      };

      return await this.createAgent(agentData);
    } catch (error) {
      console.error('Error creating agent from template:', error);
      throw new Error(`Failed to create agent from template: ${error}`);
    }
  }

  private createToolsFromRecommendations(recommendedTools: string[]): AgentTool[] {
    const toolMap: Record<string, AgentTool> = {
      web_search: {
        id: uuidv4(),
        name: 'Web Search',
        description: 'Search the web for current information',
        type: AgentToolType.WEB_SEARCH,
        config: {},
        enabled: true
      },
      calculator: {
        id: uuidv4(),
        name: 'Calculator',
        description: 'Perform mathematical calculations',
        type: AgentToolType.CALCULATOR,
        config: {},
        enabled: true
      }
    };

    return recommendedTools
      .map(toolName => toolMap[toolName])
      .filter(tool => tool !== undefined);
  }

  public async incrementUsageCount(agentId: string): Promise<void> {
    try {
      await AgentModel.findByIdAndUpdate(
        agentId,
        { $inc: { usageCount: 1 } }
      );
    } catch (error) {
      console.error(`Error incrementing usage count for agent ${agentId}:`, error);
    }
  }

  public async updateAgentRating(agentId: string, rating: number): Promise<void> {
    try {
      await AgentModel.findByIdAndUpdate(
        agentId,
        { $set: { rating } }
      );
    } catch (error) {
      console.error(`Error updating rating for agent ${agentId}:`, error);
    }
  }

  public async getPopularAgents(limit: number = 10): Promise<Agent[]> {
    try {
      const agents = await AgentModel.find({ isPublic: true })
        .sort({ usageCount: -1, rating: -1 })
        .limit(limit)
        .exec();
      
      return agents;
    } catch (error) {
      console.error('Error fetching popular agents:', error);
      return [];
    }
  }

  public async searchAgents(query: string, userId?: string): Promise<Agent[]> {
    try {
      let filter: any = {
        $or: [
          { name: { $regex: query, $options: 'i' } },
          { description: { $regex: query, $options: 'i' } },
          { tags: { $in: [new RegExp(query, 'i')] } }
        ]
      };

      if (userId) {
        filter.$or = [
          { createdBy: userId },
          { isPublic: true }
        ];
      }

      const agents = await AgentModel.find(filter).exec();
      return agents;
    } catch (error) {
      console.error('Error searching agents:', error);
      return [];
    }
  }
}

// Export singleton instance
export const agentService = new AgentService(); 